index.js

```diff
@@ -155,6 +156,65 @@ class WebTorrent extends EventEmitter {
     }
   }
 
+  /**
+   * Accepts a path to a service worker file to register, or an existing service worker registration
+   * [navigator.serviceWorker.controller] which must be activated, creates a file server for streamed
+   * file rendering to use.
+   *
+   * @param  {string|ServiceWorker} registration
+   * @param {function=} cb
+   * @return {null}
+   */
+  async registerWorker (registration, cb = () => {}) {
+    if (!('serviceWorker' in navigator)) throw new Error('Service workers not supported')
+
+    if (typeof registration === 'string') {
+      if (navigator.serviceWorker.controller !== null) throw new Error('A worker is already registered')
+      await navigator.serviceWorker.register(registration, {
+        scope: window.location.pathname.substr(0, window.location.pathname.lastIndexOf('/') + 1)
+      }).catch(e => {
+        // when loaded by redirects, workers sometimes enter an invalid state, this always happens when loaded by a protocol handler
+        // if (String(e) === 'InvalidStateError: Failed to register a ServiceWorker: The document is in an invalid state.') {
+        //   window.location.reload() // weird workaround for a weird bug
+        // }
+        throw e
+      })
+    } else {
+      if (!(registration instanceof ServiceWorker)) throw new Error('Invalid worker registration')
+      if (registration.state !== 'activated') throw new Error('Worker isn\'t activated')
+    }
+
+    this.serviceWorker = registration instanceof ServiceWorker ? registration : navigator.serviceWorker.controller
+
```

Comment:
I think the utility here is all in the parts below... and hiding it inside a fn to register a sw will prevent people who already have afunctioning sw from hooking the sw message logic.   I think a well-defined message protocol for page<-> sw would be a better approach with addins to both hook sw fetch events and page message events that people could register themselves.

wt could automatically look for sw parents and by default use the below implementation, or could be provided with an alternate compliant handler for devs who want to change behavior.
or... i've seen this be a common way to get a certain default config
`{ option: true | () => {}}`  true for default impl, fn to override.   not specified, doesn't do anything.

I don't think trying to register by default would create a problem...  you could invert it and take an impl if it's provided and not register if it's === false

I still do not think wt should be in the business of managing sws for webpages.

Reply:
yeah, the registering service workers part is kinda meh, but I do want to let users just run WT's function to register this service worker, so they consciously know what's happening with the service worker and what state it's in, I could just assume "hey this page has a service worker, lets trust it has the built-in WT functionality" or like you said create a message channel checking if said functionality exists, but I honestly was too lazy to do that, so I let users just pass a registered service worker as an option which bypasses the registration

Reply:
I changed this, and removed the service worker registration, and now only allow a sw object to be passed in, because I still want to force the user to consciously make the decision that they "activate" the sw functionality in wt, instead of automatically detecting when a service worker is registered, this entire thing could be worked around by implementing functionality which checks if the sw has some custom wt functionality it otherwise wouldn't have, so I'm gonna keep this open for now

Reply:
IMO the user should be in the control of what SW webtorrent uses, especially if there are multiple, so I wont do auto-detection